### DISCUSSION TITLE
create custom endpoint.

### DISCUSSION THREAD

**Discussion by ajaytank1991** [plausible/analytics#4543] [Self-Hosted Support]
Hello

I want to create custom endpoint. 
How can I do it in plausible CE? 

my requirement is that,

I have a page for ex domain.com/contact-us I am changed the the page slug to domain.com/contact

In state domain.com/contact consider as new page. 

I want move old slug data to new slug so If I checked page state it would be same. 

But I handle it. My project is in react js/ nodejs 

I will trigger that end point in my project when slug change.

**Comment by ruslandoga** [comment]
👋 @ajaytank1991 

There is no functionality for this at the moment. ~~You might be able to do that with a reverse proxy but that's something you would need to research on your own.~~

**Reply by ajaytank1991** [reply]
👋 @ruslandoga 

Can you please give me basic hint for it? 

How can I use reverse proxy for create custom endpoint? 

It will be helpful to me if you share some snippet or documentation.

Also can you can please let me know how to connect plausible CE database in my node server? 

So I will manage my customization from my end. 

Thanks

**Reply by ruslandoga** [reply]
That's something you would need to research on your own :) ChatGPT might be helpful.

**Reply by ajaytank1991** [reply]
I am asking here, because of I can not found any solutions. 

It will be helpful to me if you can able to share some snippet or documentation.

Also can you can please let me know how to connect plausible CE database in my node server?

Thank you.

**Reply by ruslandoga** [reply]
Ah, sorry. I misunderstood what you were trying to do. I thought you wanted to add a `/contact` page on a Plausible CE instance.

To update the data for one of the pages you can run the following commands on your [ClickHouse database.](https://github.com/plausible/community-edition/blob/4e949e84f787702835ec0654c7f663dc2fb61dbe/docker-compose.yml#L12-L23)

```console
$ cd hosting # or wherever you cloned this repo
$ docker compose exec plausible_events_db clickhouse client --database plausible_events_db
```
```sql
-- you can further limit the updates to a site_id, etc.
:) alter table events_v2 update pathname = 'contact' where pathname = 'contact-us';
:) alter table sessions_v2 update entry_page = 'contact' where entry_page = 'contact-us';
:) alter table sessions_v2 update exit_page = 'contact' where exit_page = 'contact-us';
```

**Reply by ruslandoga** [reply]
> Also can you can please let me know how to connect plausible CE database in my node server?

I assume you mean [ClickHouse,](https://github.com/plausible/community-edition/blob/4e949e84f787702835ec0654c7f663dc2fb61dbe/docker-compose.yml#L12-L23) where the event data is stored.

You can use any of these clients for Node: https://clickhouse.com/docs/en/interfaces/third-party/client-libraries#nodejs

Note that to be able to connect to `plausible_events_db`, you would need to either run your container within the same Docker network or publish 8123 and/or 9000 ports on `plausible_events_db`, e.g.:

```diff
  plausible_events_db:
    image: clickhouse/clickhouse-server:24.3.3.102-alpine
    restart: always
    volumes:
      - event-data:/var/lib/clickhouse
      - event-logs:/var/log/clickhouse-server
      - ./clickhouse/clickhouse-config.xml:/etc/clickhouse-server/config.d/logging.xml:ro
      - ./clickhouse/clickhouse-user-config.xml:/etc/clickhouse-server/users.d/logging.xml:ro
    ulimits:
      nofile:
        soft: 262144
        hard: 262144
+   ports:
+     - 127.0.0.1:8123:8123 # http interface
+     - 127.0.0.1:9000:9000 # native interface
```
This would allow accessing the database from the same host.

**Reply by ajaytank1991** [reply]
Thank you. 

Let me try it.

**Reply by ajaytank1991** [reply]
👋 @ruslandoga 

Thanks for details. I was really helpful to me. 

given snippet is working fine with `plausible_events_db`  But It is not woks with `plausible_db`.

I have did this changes 

```
plausible_db:
    # Plausible v2.1.1 was tested against PostgreSQL versions 15 and 16
    image: postgres:16-alpine
    restart: always
    volumes:
      - db-data:/var/lib/postgresql/data
    environment:
      - POSTGRES_PASSWORD=postgres
    ports:
      - "5432:5432"
```

and this code for node js

```
const clickhouse = new ClickHouse({
	url: 'http://localhost:5432',
	debug: false,
	basicAuth: null,
	isUseGzip: false,
	trimQuery: false,
	usePost: false,
	format: "json", // "json" || "csv" || "tsv"
	raw: false,
    config: {
		session_timeout                         : 6000,
		output_format_json_quote_64bit_integers : 0,
		enable_http_compression                 : 0,
		database                                : 'plausible_db',
	},
});
```

When I verify the status using curl
curl http://localhost:8123 
it's return 200 OK - it is working fine

curl http://localhost:5432
it's return 500 bad request.

Do you have any idea what I missed here?

It will be very helpful to me if you can assist me.

Thank you.

**Reply by ruslandoga** [reply]
localhost:8123 is ClickHouse, localhost:5432 is PostgreSQL. Please do use ChatGPT or other similar tools for the rest. This is no longer related to Plausible :) Good luck!

**Reply by ajaytank1991** [reply]
Ooh Okay, 

Can we connect `plausible_db` database in node project? 

If yes, how we can do it? 

If no, Can we create custom endpoint(I mean API) in Plausible CE? I will trigger that api where I want in node project. 

Or should we user our custom PostgreSQL database?

Thank you.

**Reply by ajaytank1991** [reply]
>> localhost:8123 is ClickHouse, localhost:5432 is PostgreSQL.
I got your point. 

Can we connect `plausible_db` database in node project?

If yes, how we can do it? it will helpful to me is you can assist me. 

If no, Can we create custom endpoint(I mean API) in Plausible CE? I will trigger that api where I want in node project.
Or should we use our custom PostgreSQL database for `plausible_db ` in Plausible CE?

Thank you
